Translate Chapter 05 to PTBR (#10)
JuliaIntro/JuliaIntroBR.jl: book/src/chap05.asciidoc

[[chap05]]
== Condicionais e Recursão

> O principal tópico deste capítulo é o comando +if+, que executa diferentes códigos dependendo do estado do programa. Mas antes gostaria de introduzir dois novos operadores: divisão inteira e módulo.

Reviewer: ```suggestion
O principal tópico deste capítulo é o comando +if+, que executa diferentes códigos dependendo do estado do programa. Mas antes gostaríamos de introduzir dois novos operadores: divisão inteira e módulo.
```

=== Divisão Inteira e Módulo

> O operador de _divisão inteira_, +÷+ (*+\div TAB+*), divide dois números e arredonda pra baixo para um inteiro. Por exemplo, suponha que o tempo de duração de um filme seja de 105 minutos. Talvez voc"e queira saber o quanto isso equivale em horas. A divisão convencional retorna um número em ponto-flutuante:

Reviewer: ```suggestion
O operador de _divisão inteira_, +÷+ (*+\div TAB+*), divide dois números e arredonda pra baixo para um inteiro. Por exemplo, suponha que o tempo de duração de um filme seja de 105 minutos. Talvez você queira saber o quanto isso equivale em horas. A divisão convencional retorna um número em ponto-flutuante:
```

[[chap05]]
== Condicionais e Recursão

> O principal tópico deste capítulo é a atribuição +if+, que executa diferentes códigos dependendo do critério do programa. Mas antes eu quero introduzir dois novos operadores: Divisão Inteira e módulo.

Reviewer: ```suggestion
O principal tópico deste capítulo é o comando +if+, que executa diferentes códigos dependendo do estado do programa. Mas antes vamos introduzir dois novos operadores: divisão inteira e módulo.
```

> O principal tópico deste capítulo é a atribuição +if+, da qual executa diferentes códigos dependendo do critério do programa. Mas antes eu quero introduzir dois novos operadores: Divisão Inteira e módulo.

Reviewer 2: ```suggestion
O principal tópico deste capítulo é a atribuição +if+, que executa diferentes códigos dependendo do critério do programa. Mas antes eu quero introduzir dois novos operadores: Divisão Inteira e módulo.
```

> === Divisão Floor e Módulo

Reviewer 2: Divisão inteira para todos os subsequentes "Divisão floor"

> O operador _divisão floor_, +÷+ (*+\div TAB+*), divide dois números e arredonda pra baixo para um inteiro. Por exemplo, suponha que o tempo de execução de um filme seja de 105 minutos. Você talvez saiba o quanto equivale em horas. Convencionalmente retorna um número em ponto-flutuante:

Reviewer 2: "O operador de _divisão inteira_ ..." e "A divisão convencional retorna ..."

> O operador de _Divisão Inteira_, +÷+ (*+\div TAB+*), divide dois números e arredonda pra baixo para um inteiro. Por exemplo, suponha que o tempo de execução de um filme seja de 105 minutos. Você talvez saiba o quanto equivale em horas. A divisão convencional retorna um número em ponto-flutuante:

Reviewer 2: ```suggestion
O operador de _Divisão Inteira_, +÷+ (*+\div TAB+*), divide dois números e arredonda pra baixo para um inteiro. Por exemplo, suponha que o tempo de duração de um filme seja de 105 minutos. Você talvez saiba o quanto isso equivale em horas. A divisão convencional retorna um número em ponto-flutuante:
```

> O operador de _Divisão Inteira_, +÷+ (*+\div TAB+*), divide dois números e arredonda pra baixo para um inteiro. Por exemplo, suponha que o tempo de duração de um filme seja de 105 minutos. Você talvez saiba o quanto isso equivale em horas. A divisão convencional retorna um número em ponto-flutuante:

Reviewer: ```suggestion
O operador de _divisão inteira_, +÷+ (*+\div TAB+*), divide dois números e arredonda pra baixo para um inteiro. Por exemplo, suponha que o tempo de duração de um filme seja de 105 minutos. Você talvez queira saber o quanto isso equivale em horas. A divisão convencional retorna um número em ponto-flutuante:
```

> (((operador Divisão Inteira)))((("operador", "Base", "÷", see="operador Divisão Inteira")))((("÷", see="operador Divisão Inteira")))

Reviewer 2: ```suggestion
(((operador divisão inteira)))((("operador", "Base", "÷", see="operador divisão inteira")))((("÷", see="operador divisão inteira")))
```

[source,@julia-repl-test chap05]
----
julia> minutos = 105
105
julia> minutos / 60
1.75
----

> Porém normalmente não escrevemos horas com pontos decimais. Divisão Inteira retorna um número inteiro de horas, arredondando para baixo:

Reviewer 2: ```suggestion
Porém normalmente não escrevemos horas com pontos decimais. A divisão inteira retorna um número inteiro de horas, arredondando para baixo:
```

[source,@julia-repl-test chap05]
----

> julia> houras = minutos ÷ 60

Reviewer 2: horas

1
----

> Para obter o restante, você pode subtrair uma hora em minutos:

Reviewer 2: ```suggestion
Para obter o resto, você pode subtrair uma hora em minutos:
```

> Para obter um restante, você pode subtrair uma hora em minutos:

Reviewer 2: "Para obter o restante ... "

[source,@julia-repl-test chap05]
----

> julia> restante = minutos - houras * 60

Reviewer 2: ```suggestion
julia> resto = minutos - horas * 60
```

45
----

> Uma alternativa é usar o _operador módulo_, +%+, que divide dois números e retorna o restante.

Reviewer 2: "... e retorna o resto"

[TIP]
====

> O operador módulo é mais útil do que parece. Por exemplo, você pode verificar se um número é disívell por outro — se +x % y+ é zero, então +x+ é divisível por +y+.

Reviewer 2: divisível

> Além disso, você pode extrair o dígito ou dígitos mais à direita de um número. Por exemplo, +x % 10+ gera o dígito mais à direita de um inteiro +x+ (na base 10). Da mesma forma +x % 100+ produtos os últimos dois dígitos.

Reviewer 2: "... produz os últimos dois dígitos"

====

=== Expressões Booleanas

Uma _expressão booleana_ é uma expressão que é ou verdadeira ou falsa.  Os seguintes exemplos usam o operador +==+, que compara dois operandos e gera +true+ se forem iguais e +false+ footnote:[_True_ e _False_ são comunente utilizadas nas linguagens de programação e são traduzidas, respectivamente como _verdadeiro_ e _falso_]

> (((boolean expression)))(((true)))(((false)))

Reviewer 2: "expressão booleana" no parênteses

> 

Reviewer: Passaram essas linhas acima e não consigo commitar direto.

[TIP]
====

> O operador módulo é mais útil do que parece. Por exemplo, você pode verificar se um número é divisível por outro — se +x % y+ é zero, então +x+ é divisível por +y+.

Reviewer: ```suggestion
O operador módulo é mais útil do que parece. Por exemplo, você pode verificar se um número é divisível por outro—se +x % y+ é zero, então +x+ é divisível por +y+.
```

Além disso, você pode extrair o dígito ou dígitos mais à direita de um número. Por exemplo, +x % 10+ gera o dígito mais à direita de um inteiro +x+ (na base 10). Da mesma forma +x % 100+ produz os últimos dois dígitos.
====

=== Expressões Booleanas

> Uma _expressão booleana_ é uma expressão que é ou verdadeira ou falsa.  Os seguintes exemplos usam o operador +==+, que compara dois operandos e gera +true+ se forem iguais e +false+ footnote:[_True_ e _False_ são comunente utilizadas nas linguagens de programação e são traduzidas, respectivamente como _verdadeiro_ e _falso_]

Reviewer: ```suggestion
Uma _expressão booleana_ é uma expressão que é ou verdadeira ou falsa.  Os seguintes exemplos usam o operador +==+, que compara dois operandos e gera +true+ se forem iguais e +false+--do inglês, verdadeiro e falso.
```

====

=== Operadores Lógicos

Existem três _Operadores Lógicos_: +&&+ (e), +||+ (ou) e +!+ (não). A semântica (significado) destes operadores são similiares aos seus significados em Português. Por exemplo, +x > 0 && x < 10+ é verdadeiro se e somente se +x+ é maior que 0 _e_ menor que +10+.
(((operador lógico)))(((&&)))(((||)))(((!)))

+n % 2 == 0 || n % 3 == 0+ é verdadeiro se _uma ou ambas_ condições são verdadeiras, isto é, se o número é divisível por 2 _ou_ por 3.

Ambos +&&+ e +||+ associam à direita, mas +&&+ tem maior precedência que +||+.

Por fim, o operador +!+ nega uma expressão booleana, então +!(x > y)+ é verdadeira se +x > y+ é falsa, isto é, se +x+ é menor ou igual a +y+.

=== Execução Condicional

Para escrever programas úteis, quase sempre precisamos verificar as condições e alterar o comportamento de acordo com o programa. _Declarações Condicionais_ nos fornece essa habilidade. A forma mais simples é a atribuição _se_ (_if_ em Inglês).
(((atribuição condicional)))(((if)))((("palavra-chave", "se", see="se")))(((atribuição se)))((("atribuição", "se", see="atribuição se")))

[source,julia]
----
if x > 0
    println("x é positivo")
end
----

> A expressão booleana depois de +if+ é chamada de _condição_. Se é verdadeira, então a atribuição identada é executada. Se não, nada acontece. (((condição)))

Reviewer 2: " ... atribuição indentada ... "

A atribuição _se_ possui a mesma estrutura das definições de funções: Um cabeçalho seguido de um corpo terminado com a palavra-chave _end_. Declarações como essa são chamadas de _declarações compostas_.

> (((declrações compostas)))(((end)))(((fim)))

Reviewer 2: declarações e remova o "(((fim)))"

Não há limites para o número de declarações que podem aparecer no corpo. Ocasionalmente, é útil ter um corpo sem declarações (geralmente como detentor de um lugar para o código que você ainda não escreveu).

[source,julia]
----
if x < 0
    # TODO: precisa lidar com valores negativos!
end
----

=== Execuções Alternativas

Uma segunda maneira de usar a atribuição _se_ é a "execução alternativa", da qual há duas possibilidades e a condição determina qual delas deverá ser executada. A sintaxe é a seguinte:

> (((execução alternativa)))(((se não)))((("palavra-chame", "senão", see="se não")))

Reviewer 2: "palavra-chame" -> "palavra-chave" e "senão" -> "se não"

[source,julia]
----
if x % 2 == 0
    println("x é par")
else
    println("x é ímpar")
end
----

> Se o resto quando +x+ é dividido por 2 é 0, então sabemos que +x+ é par, e o programa irá exibir a mensagem apropriada. Se a condição for falsa, o segundo conjunto de declarações é executado. Desde que a condição seja verdadeira ou falsa, exatamente uma das alternativas irá ser executada. Essas alternativas são chamadas de _ramos_ (_branches_ em inglês), porque são ramos de um fluxo de execução.

Reviewer 2: acho que facilita pro leitor algo como "Se o resto da divisão de x por 2 é 0, ... "

(((ramos)))

=== Condicionais Encadeadas

> Algumas vezes há mais do que duas possibilidades e precisamos de mais que dois ramos. Uma maneira de expressar um cálculo como esse é através das _condicionais encadeadas_:

Reviewer 2: Concordo que a tradução de "computation" é "cálculo" mas acho que seria melhor substituir por algo mais próximo de "comando" ou  "operação" , dado o contexto

(((condicionais encadeadas)))(((se então)))((("palavra-chave", "se então", see="se então")))

[source,julia]
----
if x < y
    println("x é menor que y")
elseif x > y
    println("x é maior que y")
else
    println("x e y são iguais")
end
----

Novamente, exatamente um dos ramos será executado. Não há limites para o número de declarações +else if+. Se existir uma cláusula +else+, essa deve estar no final, mas não precisa haver uma.

[source,julia]
----
if escolha == "a"
    draw_a()
elseif escolha == "b"
    draw_b()

> elseif esculha == "c"

Reviewer 2: escolha

    draw_c()
end
----

Cada condição é checada em ordem. Se a primeira for falsa, a próxima é checada e assim por diante. Se uma delas é verdadeira, o ramo correspondente é executado e a atribuição é encerrada. Se mais de uma condição é verdadeira, apenas o primeiro ramo verdadeiro é executado.

=== Condicionais Aninhadas

Uma condicional também pode ser aninhada com outra. Poderíamos ter escrito no exemplo da seção anterior da seguinte forma:
(((condicionais aninhadas)))

[source,julia]
----
if x == y
    println("x e y são iguais")
else
    if x < y
        println("x é menor que y")
    else
        println("x é maior que y")
    end
end
----

O condicional externo contém dois ramos. O primeiro ramo contém uma atribuição simples. O segundo ramo contém outra atribuição +if+, da qual possui dois ramos inserida nela. Esses dois ramos são declarações simples, embora também possam ter sido declarações condicionais.

> Embora a indentação não obrigatório das declarações torne a estrutura aparente, _condicionais aninhadas_ tornam-se difíceis de ler muito rapidamente. É uma boa ideia evitá-los quando puder.

Reviewer 2: "obrigatório" -> "obrigatória"

(((indentação)))

Operadores lógicos geralmente fornecem uma maneira de simplificar instruções condicionais aninhadas. Por exemplo, podemos reescrever o seguinte código usando uma única condicional:

[source,julia]
----
if 0 < x
    if x < 10
        println("x é um número positivo de um dígito.")
    end
end
----

A atribuição +print+ executa somente se passarmos pelos dois condicionais, para que possamos obter o mesmo efeito com operador +&&+:

[source,julia]
----
if 0 < x && x < 10
    println("x é um número positivo de um dígito.")
end
----

> Para esse tipo de condição, Julia fornece uma sintaxe mais concisa:

Reviewer 2: "... , O Julia fornece ... "

[source,julia]
----
if 0 < x < 10
    println("x é um número positivo de um dígito.")
end
----

[[recursion]]
=== Recursão

> É permitido que uma função chame outra; também é permitido uma função se chamar. Pode não ser óbvio por que isso é uma coisa boa, mas acaba sendo uma das coisas mais mágicas que um programa pode fazer. Por exemplo, observe a seguinte função:

Reviewer 2: acho que fica mais claro "se chamar" -> "chamar a si mesmo"

> (((contagem regressiva)))((("função", "definido pelo programador", "contagem regressiva", see="contagem")))

Reviewer 2: "contagem" -> "contagem regressiva"

> (((contador regressivo)))((("função", "programmer-defined", "contador regressivo", see="contador")))

Reviewer 2: "definido pelo programador"

[source,@julia-setup chap05]
----
function contagem_regressiva(n)
    if n ≤ 0
        println("Vai!")
    else
        print(n, " ")
        contagem_regressiva(n-1)
    end
end
----

> Se +n+ é 0 ou negativo, será exibido a palavra, +"Vai!"+ de outra forma, exibirá +n+ e executará a função chamada +contagem_regressiva+ — a própria — passando +n-1+ como argumento.

Reviewer 2: " de outra forma, exibirá" -> " por outro lado, a função exibirá" e " a própria " -> "ela mesma"

Bool
----

O operador +==+ é um operador de relação; os outros são:
(((operador de relação)))(((==)))((("operador", "Base", "==", see="==")))

[source,julia]
----
      x != y               # x não é igual a y
      x ≠ y                # (\ne TAB)
      x > y                # x é maior que y
      x < y                # x é menor que y
      x >= y               # x é maior ou igual a y
      x ≥ y                # (\ge TAB)
      x <= y               # x é menor ou igual a y
      x ≤ y                # (\le TAB)
----

[WARNING]
====

> Embora essas operações sejam provavelmente familiares para você, os símbolos Julia são diferentes dos símbolos matemáticos. Um erro comum é usar um único sinal de igual (+=+) em vez de um duplo sinal de igual (+==+). Lembre-se de que +=+ é um operador de atribuição e +==+ é um operador relacional. Não existe +=<+ ou +pass:[=>]+.

Reviewer: ```suggestion
Embora essas operações sejam provavelmente familiares para você, os símbolos em Julia são diferentes dos símbolos matemáticos. Um erro comum é usar um único sinal de igual (+=+) em vez de um duplo sinal de igual (+==+). Lembre-se de que +=+ é um operador de atribuição e +==+ é um operador relacional. Não existe +=<+ ou +pass:[=>]+.
```

(((≠)))((("operador", "Base", "≠", see="≠")))((("!=", see="≠")))((("operador", "Base", "!=", see="≠")))(((>)))((("operador", "Base", ">", see=">")))(((<)))((("operador", "Base", "<", see="<")))(((≥)))((("operador", "Base", "≥", see="≥")))(((">=", see="≥")))((("operador", "Base", ">=", see="≥")))(((≤)))((("operador", "Base", "≤", see="≤")))((("pass:[&lt;=]", see="≤")))((("operador", "Base", "pass:[&lt;=]", see="≤")))
====

=== Operadores Lógicos

> Existem três _Operadores Lógicos_: +&&+ (e), +||+ (ou) e +!+ (não). A semântica (significado) destes operadores são similiares aos seus significados em Português. Por exemplo, +x > 0 && x < 10+ é verdadeiro se e somente se +x+ é maior que 0 _e_ menor que +10+.

Reviewer: ```suggestion
Existem três _operadores lógicos_: +&&+ (e), +||+ (ou) e +!+ (não). A semântica (significado) destes operadores são similiares aos seus significados em Português. Por exemplo, +x > 0 && x < 10+ é verdadeiro se e somente se +x+ é maior que 0 _e_ menor que +10+.
```

(((operador lógico)))(((&&)))(((||)))(((!)))

+n % 2 == 0 || n % 3 == 0+ é verdadeiro se _uma ou ambas_ condições são verdadeiras, isto é, se o número é divisível por 2 _ou_ por 3.

Ambos +&&+ e +||+ associam à direita, mas +&&+ tem maior precedência que +||+.

Por fim, o operador +!+ nega uma expressão booleana, então +!(x > y)+ é verdadeira se +x > y+ é falsa, isto é, se +x+ é menor ou igual a +y+.

=== Execução Condicional

> Para escrever programas úteis, quase sempre precisamos verificar as condições e alterar o comportamento de acordo com o programa. _Declarações Condicionais_ nos fornece essa habilidade. A forma mais simples é a atribuição _se_ (_if_ em Inglês).

Reviewer 2: ```suggestion
Para escrever programas úteis, quase sempre precisamos verificar as condições e alterar o comportamento de acordo com o programa. _Declarações Condicionais_ nos fornecem essa habilidade. A forma mais simples é a atribuição _se_ (_if_ em Inglês).
```

> Para escrever programas úteis, quase sempre precisamos verificar as condições e alterar o comportamento de acordo com o programa. _Declarações Condicionais_ nos fornecem essa habilidade. A forma mais simples é a atribuição _se_ (_if_ em Inglês).

Reviewer: ```suggestion
Para escrever programas úteis, quase sempre precisamos verificar as condições e alterar o comportamento de acordo com o programa. _Declarações condicionais_ nos fornecem essa habilidade. A forma mais simples é o comando _if_ (da conjunção se em Inglês).
```

> (((atribuição condicional)))(((if)))((("palavra-chave", "se", see="se")))(((atribuição se)))((("atribuição", "se", see="atribuição se")))

Reviewer: ```suggestion
(((declaração condicional)))(((if)))((("palavra-chave", "if", see="if")))(((comando if)))((("comando", "if", see="comando if")))
```

[source,julia]
----
if x > 0
    println("x é positivo")
end
----

> A expressão booleana depois de +if+ é chamada de _condição_. Se é verdadeira, então a atribuição indentada é executada. Se não, nada acontece. 

Reviewer: ```suggestion
A expressão booleana depois de +if+ é chamada de _condição_. Se é verdadeira, então o comando indentado é executado. Se não, nada acontece. 
```

(((condição)))

> A atribuição _se_ possui a mesma estrutura das definições de funções: Um cabeçalho seguido de um corpo terminado com a palavra-chave _end_. Declarações como essa são chamadas de _declarações compostas_.

Reviewer: ```suggestion
Declarações _if_ possuem a mesma estrutura das definições de funções: Um cabeçalho seguido de um corpo terminado com a palavra-chave _end_. Declarações como essa são chamadas de _declarações compostas_.
```

(((declarações compostas)))(((end)))

Não há limites para o número de declarações que podem aparecer no corpo. Ocasionalmente, é útil ter um corpo sem declarações (geralmente como detentor de um lugar para o código que você ainda não escreveu).

[source,julia]
----
if x < 0

>     # TODO: precisa lidar com valores negativos!

Reviewer: ```suggestion
    # LEMBRETE: precisa lidar com valores negativos!
```

end
----

=== Execuções Alternativas

> Uma segunda maneira de usar a atribuição _se_ é a "execução alternativa", da qual há duas possibilidades e a condição determina qual delas deverá ser executada. A sintaxe é a seguinte:

Reviewer: ```suggestion
Uma segunda maneira de usar o comando _if_ é a "execução alternativa", da qual há duas possibilidades e a condição determina qual delas deverá ser executada. A sintaxe é a seguinte:
```

> (((execução alternativa)))(((se não)))((("palavra-chave", "se não", see="se não")))

Reviewer: ```suggestion
(((execução alternativa)))(((else)))((("palavra-chave", "else", see="else")))
```

[source,julia]
----
if x % 2 == 0
    println("x é par")
else
    println("x é ímpar")
end
----

Se o resto da divisão de +x+ por 2 é 0, então sabemos que +x+ é par, e o programa irá exibir a mensagem apropriada. Se a condição for falsa, o segundo conjunto de declarações é executado. Desde que a condição seja verdadeira ou falsa, exatamente uma das alternativas irá ser executada. Essas alternativas são chamadas de _ramos_ (_branches_ em inglês), porque são ramos de um fluxo de execução.
(((ramos)))

=== Condicionais Encadeadas

Algumas vezes há mais do que duas possibilidades e precisamos de mais que dois ramos. Uma maneira de expressar um comando como esse é através das _condicionais encadeadas_:

> (((condicionais encadeadas)))(((se então)))((("palavra-chave", "se então", see="se então")))

Reviewer 2: ```suggestion
(((condicionais encadeadas)))(((elseif)))((("palavra-chave", "elseif", see="elseif")))
```

[source,julia]
----
if x < y
    println("x é menor que y")
elseif x > y
    println("x é maior que y")
else
    println("x e y são iguais")
end
----

> Novamente, exatamente um dos ramos será executado. Não há limites para o número de declarações +else if+. Se existir uma cláusula +else+, essa deve estar no final, mas não precisa haver uma.

Reviewer 2: ```suggestion
Novamente, exatamente um dos ramos será executado. Não há limites para o número de declarações +elseif+. Se existir uma cláusula +else+, essa deve estar no final, mas não precisa haver uma.
```

[source,julia]
----
if escolha == "a"
    draw_a()
elseif escolha == "b"
    draw_b()
elseif escolha == "c"
    draw_c()

> end

Reviewer: Mudar `draw` para `desenha`.

----

Cada condição é checada em ordem. Se a primeira for falsa, a próxima é checada e assim por diante. Se uma delas é verdadeira, o ramo correspondente é executado e a atribuição é encerrada. Se mais de uma condição é verdadeira, apenas o primeiro ramo verdadeiro é executado.

=== Condicionais Aninhadas

Uma condicional também pode ser aninhada com outra. Poderíamos ter escrito no exemplo da seção anterior da seguinte forma:
(((condicionais aninhadas)))

[source,julia]
----
if x == y
    println("x e y são iguais")
else
    if x < y
        println("x é menor que y")
    else
        println("x é maior que y")
    end
end
----

> O condicional externo contém dois ramos. O primeiro ramo contém uma atribuição simples. O segundo ramo contém outra atribuição +if+, da qual possui dois ramos inserida nela. Esses dois ramos são declarações simples, embora também possam ter sido declarações condicionais.

Reviewer: ```suggestion
O condicional externo contém dois ramos. O primeiro ramo contém uma declaração simples. O segundo ramo contém outro condicional +if+, que possui dois ramos inseridos nele. Esses dois ramos são declarações simples, embora também possam ter sido declarações condicionais.
```

Embora a indentação não obrigatória das declarações torne a estrutura aparente, _condicionais aninhadas_ tornam-se difíceis de ler muito rapidamente. É uma boa ideia evitá-los quando puder.
(((indentação)))

Operadores lógicos geralmente fornecem uma maneira de simplificar instruções condicionais aninhadas. Por exemplo, podemos reescrever o seguinte código usando uma única condicional:

[source,julia]
----
if 0 < x
    if x < 10
        println("x é um número positivo de um dígito.")
    end
end
----

> A atribuição +print+ executa somente se passarmos pelos dois condicionais, para que possamos obter o mesmo efeito com operador +&&+:

Reviewer: ```suggestion
O comando +print+ executa somente se passarmos pelos dois condicionais, para que possamos obter o mesmo efeito com operador +&&+:
```

[source,julia]
----
if 0 < x && x < 10
    println("x é um número positivo de um dígito.")
end
----

Para esse tipo de condição, o Julia fornece uma sintaxe mais concisa:

[source,julia]
----
if 0 < x < 10
    println("x é um número positivo de um dígito.")
end
----

[[recursion]]
=== Recursão

É permitido que uma função chame outra; também é permitido uma função chamar si mesmo. Pode não ser óbvio por que isso é uma coisa boa, mas acaba sendo uma das coisas mais mágicas que um programa pode fazer. Por exemplo, observe a seguinte função:

> (((contagem regressiva)))((("função", "definido pelo programador", "contagem regressiva", see="contagem")))

Reviewer 2: ```suggestion
(((contagem regressiva)))((("função", "definido pelo programador", "contagem regressiva", see="contagem regressiva")))
```

[source,@julia-setup chap05]
----
function contagem_regressiva(n)
    if n ≤ 0
        println("Vai!")
    else
        print(n, " ")
        contagem_regressiva(n-1)
    end
end
----

> Se +n+ é 0 ou negativo, será exibido a palavra, +"Vai!"+, por outro lado, exibirá +n+ e executará a função chamada +contagem_regressiva+ — ela mesma — passando +n-1+ como argumento.

Reviewer: ```suggestion
Se +n+ é 0 ou negativo, será exibido a palavra +"Vai!"+. Caso contrário, a função exibe +n+ e chama uma função chamada +contagem_regressiva+—ela mesma— passando +n-1+ como argumento.
```

O que acontece se chamarmos uma função como essa?

[source,@julia-repl-test chap05]
----
julia> contagem_regressiva(3)
3 2 1 Vai!
----

> A execução de +contagem_regressiva+ começa com +n = 3+, e desde que +n+ seja maior que 0, terá como saída o valor 3, e depois executa ela mesma...

Reviewer: ```suggestion
A execução de +contagem_regressiva+ começa com +n = 3+, e como +n+ é maior que 0, terá como saída o valor 3, e depois executa ela mesma...
```

> pass:[&#8193;]A execução de +contagem_regressiva+ começa com +n = 2+, e desde que +n+ seja maior que 0, pass:[<br/>&#8193;&#8193;]terá como saída o valor 2, e depois executa ela mesma ...

Reviewer: ```suggestion
pass:[&#8193;]A execução de +contagem_regressiva+ começa com +n = 2+, e como +n+ é maior que 0, pass:[<br/>&#8193;&#8193;]terá como saída o valor 2, e depois executa ela mesma ...
```

> pass:[&#8193;&#8193;]A execução de +contagem_regressiva+ começa com +n = 1+, e desde que +n+ seja maior que 0, pass:[<br/>&#8193;&#8193;&#8193;]terá como saída o valor 1, e depois executa ela mesma ...

Reviewer: ```suggestion
pass:[&#8193;&#8193;]A execução de +contagem_regressiva+ começa com +n = 1+, e como  +n+ é maior que 0, pass:[<br/>&#8193;&#8193;&#8193;]terá como saída o valor 1, e depois executa ela mesma ...
```

> pass:[&#8193;&#8193;&#8193;]A execução de +contagem_regressiva+ começa com +n = 0+, e desde que +n+ seja maior que pass:[<br/>&#8193;&#8193;&#8193;&#8193;]0, terá como saída uma palavra, +"Vai!"+ e depois retorna.

Reviewer: ```suggestion
pass:[&#8193;&#8193;&#8193;]A execução de +contagem_regressiva+ começa com +n = 0+, e como +n+ não é maior que pass:[<br/>&#8193;&#8193;&#8193;&#8193;]0, terá como saída uma palavra, +"Vai!"+ e depois retorna.
```

pass:[&#8193;&#8193;]A contagem regressiva que obteve +n = 1+ retorna.

pass:[&#8193;]A contagem regressiva que obteve +n = 2+ retorna.

A contagem regressiva que obteve +n = 3+ retorna.

E então você voltará para +Main+.

Uma função que chama a si mesma é chamada de _recursiva_; o processo de executar ela é chamada de _recursão_.
(((função recursiva)))(((recursão)))

Um outro exemplo, podemos escrever uma função que imprime uma string latexmath:[\(n\)] vezes.

> (((printn)))((("função", "definido pelo programador", "printn", see="printn")))

Reviewer 2: ```suggestion
(((imprima_n)))((("função", "definido pelo programador", "imprima_n", see="imprima_n")))
```

+n % 2 == 0 || n % 3 == 0+ é verdadeiro se _uma ou ambas_ condições são verdadeiras, isto é, se o número é divisível por 2 _ou_ por 3.

Tanto +&&+ quanto +||+ associam à direita, mas +&&+ tem maior precedência que +||+.

Por fim, o operador +!+ nega uma expressão booleana, então +!(x > y)+ é verdadeira se +x > y+ é falsa, isto é, se +x+ é menor ou igual a +y+.

=== Execução Condicional

> Para escrever bons programas, quase sempre precisamos verificar as condições e alterar o comportamento de acordo com o programa. _Comandos condicionais_ nos fornecem essa habilidade. A forma mais simples é o comando _if_ (da conjunção se em Inglês).

Reviewer: ```suggestion
Para escrever programas úteis, quase sempre precisamos verificar as condições e alterar o comportamento de acordo com o programa. _Comandos condicionais_ nos fornecem essa habilidade. A forma mais simples é o comando _if_ (da conjunção se em Inglês).
```

[[recursion]]
=== Recursão

> É possível fazer com que uma função chame outra; também é possível uma função chamar si mesma. Pode não parecer óbvio por que isso é uma coisa boa, mas acaba sendo uma das coisas mais mágicas que um programa pode fazer. Por exemplo, observe a seguinte função:

Reviewer: ```suggestion
É possível fazer com que uma função chame outra; também é possível uma função chamar a si mesma. Pode não parecer óbvio por que isso é uma coisa boa, mas acaba sendo uma das coisas mais mágicas que um programa pode fazer. Por exemplo, observe a seguinte função:
```

end
----

> Se +n pass:[&lt;]= 0+ a atribuição +return+ sai da função. O fluxo de execução retorna imediatamente para o chamador e as linhas restantes da função não são executadas.

Reviewer 2: "chamador" -> "quem a chamou"

> (((return)))((("palavra-chave", "return", see="return")))(((return statement)))((("atribuição", "return", see="atribuição de retorno")))

Reviewer 2: "statement" -> "declaração"/"comando"

> O restante da função é similar a +contagem_regressiva+: Exibirá +s+ e chamará a si mesmo para exibir +s+ latexmath:[\(n-1\)] várias vezes. Portanto, o número de linhas de saída é latexmath:[\(1 + (n - 1)\)], o que soma latexmath:[\(n\)].

Reviewer 2: "Exibirá" -> "Ela exibirá"

Para exemplos simples como esse, provavelmente é mais fácil usar um loop +for+. Mas veremos exemplos mais tarde que são difíceis de escrever com um loop +for+ e fáceis de escrever com recursão; portanto, é bom começar cedo.

=== Diagramas de Pilhas para funções recursivas

Em <<stack_diagrams>>, usamos um diagrama de pilha para representar o estado de um programa durante uma chamada de função. O mesmo tipo de diagrama pode ajudar a interpretar uma função recursiva.
(((diagrama de pilha)))

Sempre que uma função é chamada, o Julia cria um quadro para conter as variáveis e parâmetros locais da função. Para uma função recursiva, pode haver mais de um quadro na pilha ao mesmo tempo.
(((quadro)))

[[fig05-1]]
.Diagrama de Pilha
image::images/fig51.svg[]

<<fig05-1>> mostra um diagrama de pilha para +contagem_regressiva+ chamada com +n = 3+.

> Como sempre, o topo para pilha é o quadro para +Main+. Está vazio porque não criamos nenhuma variável em +Main+ ou passamos algum argumentos para ela.

Reviewer 2: "o topo para pilha" -> "o topo da pilha"  
"Está vazio" -> "Ele está vazio"  
"ou passamos algum argumentos para ela" -> "nem passamos argumentos para ela"

> Os quatro quadros de +contagem_regressiva+ têm valores diferentes para o parâmetro +n+. A parte inferior da pilha, onde +n=0+, é chamada de _caso base_. Não faz uma chamada recursiva, portanto não há mais quadros.

Reviewer 2: "Não faz .. " -> "Ele não faz .. "

(((caso base)))

===== Exercício 5-1

> Como exercício, desenhe um diagrama de pilha para +printn+ chamado com +s = "Olá"+ e +n = 2+. Depois, escreva uma função chamada +do_n+ que pega um objeto de função e um número, +n+, como argumento, e que chama a função dada latexmath:[\(n\)] times.

Reviewer 2: "times" -> "vezes"

(((recursão infinita)))(((recursão)))((("função", "definido pelo programador", "recursão", see="recursão")))

[source,julia]
----
function recursão()
    recursão()
end
----

Na maioria dos ambientes de programação, um programa com recursão infinita realmente não é executado para sempre. O Julia exibe uma mensagem de erro quando a profundidade máxima de recursão é atingida:

[source,jlcon]
----
julia> recursão()
ERROR: StackOverflowError:
Stacktrace:
 [1] recurse() at ./REPL[1]:2 (repeats 80000 times)
----

Esse rastreamento de pilha é um pouco maior do que vimos no capítulo anterior. Quando o erro ocorre, existem 80000 quadros de +recursão+ na pilha!
(((rastreamento de pilha)))(((StackOverflowError)))((("Erro", "Core", "StackOverflowError", see="StackOverflowError")))

Se você encontrar uma recursão infinita por acidente, revise sua função para confirmar se há um caso base que não faz uma chamada recursiva. E se houver um caso base, verifique se você está garantindo para alcançá-lo.

=== Entradas de Teclado

Os programas que escrevemos até agora não aceitam nenhuma entrada do usuário. Eles apenas fazem a mesma coisa todas as vezes.

> Julia fornece uma função interna chamada +readline+ que interrompe o programa e aguarda o usuário digitar algo. Quando o usuário pressiona +RETURN+ ou +ENTER+, o programa é retomado e +readline+ retorna o que o usuário digitou como uma seqüência de caracteres.

Reviewer 2: sequência

[source,jlcon]
----
julia> println("Qual é a velocidade da velocidade do ar de uma andorinha sem carga? "); speed = readline()
Qual é a velocidade da velocidade do ar de uma andorinha sem carga?
42
"42"
julia> parse(Int64, speed)
42
----

Mas se o usuário digitar algo diferente de uma sequência de dígitos, você receberá um erro:
(((analisar)))

[source,jlcon]
----
julia> println("Qual é a velocidade da velocidade do ar de uma andorinha sem carga? "); velocidade = readline()
Qual é a velocidade da velocidade do ar de uma andorinha sem carga?
Como assim, uma andorinha africana ou européia?
"Como assim, uma andorinha africana ou européia?"
julia> parse(Int64, speed)
ERROR: ArgumentError: invalid base 10 digit 'C' in "Como assim, uma andorinha africana ou européia?"
[...]
----

Veremos como lidar com esse tipo de erro posteriormente.
(((Erro de Argumento)))

=== Debugando

Quando ocorre um erro de sintaxe ou de tempo de execução, a mensagem de erro contém muitas informações, mas pode ser avassaladora. As partes mais úteis são geralmente:
(((debugando)))

* Que tipo de erro foi, e

* Onde ocorreu.

> Os erros de sintaxe geralmente são fáceis de encontrar, mas existem algumas dicas. Em geral, as mensagens de erro indicam onde o problema foi descoberto, mas o erro real pode ser anterior no código, às vezes em uma linha anterior.

Reviewer 2: "mas o erro real pode ser anterior no código" -> "mas o verdadeiro erro pode estar antes no código"

O mesmo vale para erros de tempo de execução. Suponha que você esteja tentando calcular uma taxa de sinal/ruído em decibéis. A fórmula é

[latexmath]
++++
\begin{equation}
{SNR_{\mathrm{db}} = 10 \log_{10} \frac{P_{\mathrm{sinal}}}{P_{\mathrm{ruido}}}\ .}
\end{equation}
++++

Em Julia, você pode escrever algo como isto:

[source,julia]
----
potência_do_sinal = 9
potência_do_ruido = 10
razão = potência_do_sinal ÷ potência_do_ruido

> decibeis = 10 * log10(razão)

Reviewer 2: decibéis

(((operador módulo)))

expressão booleana::
Uma expressão cujo os valores são ou +true+ ou +false+.
(((expressão booleana)))

operador relacional::
Um dos operadores que compara operandos: +==+, +≠+ (+!=+), +>+, +<+, +≥+ (+>=+), e +≤+ (+pass:[&lt;=]+).
(((operador relacional)))

operador lógico::
Um dos operadores que combina expressões booleanas: +&&+ (e), +||+ (ou), e +!+ (não).
(((operador lógico)))

atribuição condicional::
Uma atribuição que controla o fluxo de execução dependendo de alguma condição.
(((atribuição condicional)))

condição::
A expressão boleana em uma atribuição condicional que determina qual ramo executará.
(((condição)))

atribuição composta::
Uma atribuição que consiste em um cabeçalho e um corpo. O corpo é terminado com a palavra-chave +end+.
(((atribuição composta)))

ramos::
Uma das sequências alternativas de atribuições em uma atribuição condicional.
(((ramos)))

atribuição encadeada::
Uma atribuição condicional com uma série de ramos alternativos.
(((atribuição encadeada)))

condicional aninhada::
Uma atribuição condicional que aparece em um dos ramos de outra atribuição condicional.
(((condicional aninhada)))

atribuição de retorno::

> Uma atribuição que faz com que uma função pare de executar imediatamente e retorne para o chamador da função.

Reviewer 2: "chamador" -> "quem chamou"

end

> ----

Reviewer: printn -> imprima_n

Reviewer: Linha 299!

> Se +n pass:[&lt;]= 0+ a atribuição +return+ sai da função. O fluxo de execução retorna imediatamente para quem a chamou e as linhas restantes da função não são executadas.

Reviewer: ```suggestion
Se +n pass:[&lt;]= 0+ o comando +return+ sai da função. O fluxo de execução retorna imediatamente para quem a chamou e as linhas restantes da função não são executadas.
```

> (((return)))((("palavra-chave", "return", see="return")))(((declaração de retorno)))((("declaração", "return", see="declaração de retorno")))

Reviewer 2: ```suggestion
(((return)))((("palavra-chave", "return", see="return")))(((declaração return)))((("declaração", "return", see="declaração return")))
```

> O restante da função é similar a +contagem_regressiva+: Exibirá +s+ e chamará a si mesmo para exibir +s+ latexmath:[\(n-1\)] várias vezes. Portanto, o número de linhas de saída é latexmath:[\(1 + (n - 1)\)], o que soma latexmath:[\(n\)].

Reviewer 2: ```suggestion
O restante da função é similar a +contagem_regressiva+: Ela exibirá +s+ e chamará a si mesmo para exibir +s+ latexmath:[\(n-1\)] várias vezes. Portanto, o número de linhas de saída é latexmath:[\(1 + (n - 1)\)], o que soma latexmath:[\(n\)].
```

> Para exemplos simples como esse, provavelmente é mais fácil usar um loop +for+. Mas veremos exemplos mais tarde que são difíceis de escrever com um loop +for+ e fáceis de escrever com recursão; portanto, é bom começar cedo.

Reviewer 2: ```suggestion
Para exemplos simples como esse, provavelmente é mais fácil usar um laço +for+. Mas veremos exemplos mais tarde que são difíceis de escrever com um laço +for+ e fáceis de escrever com recursão; portanto, é bom começar cedo.
```

> === Diagramas de Pilhas para funções recursivas

Reviewer: ```suggestion
=== Diagramas de Pilhas para Funções Recursivas
```

Em <<stack_diagrams>>, usamos um diagrama de pilha para representar o estado de um programa durante uma chamada de função. O mesmo tipo de diagrama pode ajudar a interpretar uma função recursiva.
(((diagrama de pilha)))

Sempre que uma função é chamada, o Julia cria um quadro para conter as variáveis e parâmetros locais da função. Para uma função recursiva, pode haver mais de um quadro na pilha ao mesmo tempo.
(((quadro)))

[[fig05-1]]
.Diagrama de Pilha
image::images/fig51.svg[]

<<fig05-1>> mostra um diagrama de pilha para +contagem_regressiva+ chamada com +n = 3+.

> Como sempre, o topo da pilha é o quadro para +Main+. Ele está vazio porque não criamos nenhuma variável em +Main+ ou nem passamos argumentos para ela.

Reviewer 2: ```suggestion
Como sempre, o topo da pilha é o quadro para +Main+. Ele está vazio porque não criamos nenhuma variável em +Main+ ou passamos algum argumento para ela.
```

> Os quatro quadros de +contagem_regressiva+ tem valores diferentes para o parâmetro +n+. A parte inferior da pilha, onde +n=0+, é chamada de _caso base_. Ele não faz uma chamada recursiva, portanto não há mais quadros.

Reviewer: ```suggestion
Os quatro quadros de +contagem_regressiva+ têm valores diferentes para o parâmetro +n+. A parte inferior da pilha, onde +n=0+, é chamada de _caso base_. Ele não faz uma chamada recursiva, portanto não há mais quadros.
```

> Os quatro quadros de +contagem_regressiva+ têm valores diferentes para o parâmetro +n+. A parte inferior da pilha, onde +n=0+, é chamada de _caso base_. Ele não faz uma chamada recursiva, portanto não há mais quadros.

Reviewer 2: ```suggestion
Os quatro quadros de +contagem_regressiva+ tem valores diferentes para o parâmetro +n+. A parte inferior da pilha, onde +n=0+, é chamada de _caso base_. Ele não faz uma chamada recursiva, portanto não há mais quadros.
```

(((caso base)))

===== Exercício 5-1

> Como exercício, desenhe um diagrama de pilha para +printn+ chamado com +s = "Olá"+ e +n = 2+. Depois, escreva uma função chamada +faça_n+ que pega um objeto de função e um número, +n+, como argumento, e que chama a função dada latexmath:[\(n\)] vezes.

Reviewer: ```suggestion
Como exercício, desenhe um diagrama de pilha para +imprima_n+ chamado com +s = "Olá"+ e +n = 2+. Depois, escreva uma função chamada +faça_n+ que pega um objeto de função e um número, +n+, como argumento, e que chama a função dada latexmath:[\(n\)] vezes.
```

> Como exercício, desenhe um diagrama de pilha para +printn+ chamado com +s = "Olá"+ e +n = 2+. Depois, escreva uma função chamada +do_n+ que pega um objeto de função e um número, +n+, como argumento, e que chama a função dada latexmath:[\(n\)] vezes.

Reviewer 2: ```suggestion
Como exercício, desenhe um diagrama de pilha para +printn+ chamado com +s = "Olá"+ e +n = 2+. Depois, escreva uma função chamada +faça_n+ que pega um objeto de função e um número, +n+, como argumento, e que chama a função dada latexmath:[\(n\)] vezes.
```

=== Recursão Infinita

> Se uma recursão nunca atinge o caso base, ele continua fazendo chamadas recursivas para sempre e o programa nunca termina. Isso é conhecido como _recursão infinita_, e geralmente isso não é uma boa ideia. Aqui está um programa pequeno com uma recursão infinita:

Reviewer 2: ```suggestion
Se uma recursão nunca atinge o caso base, ela continua fazendo chamadas recursivas para sempre e o programa nunca termina. Isso é conhecido como _recursão infinita_, e geralmente isso não é uma boa ideia. Aqui está um programa pequeno com uma recursão infinita:
```

(((recursão infinita)))(((recursão)))((("função", "definido pelo programador", "recursão", see="recursão")))

[source,julia]
----
function recursão()
    recursão()
end
----

Na maioria dos ambientes de programação, um programa com recursão infinita realmente não é executado para sempre. O Julia exibe uma mensagem de erro quando a profundidade máxima de recursão é atingida:

[source,jlcon]
----
julia> recursão()
ERROR: StackOverflowError:
Stacktrace:
 [1] recurse() at ./REPL[1]:2 (repeats 80000 times)
----

Esse rastreamento de pilha é um pouco maior do que vimos no capítulo anterior. Quando o erro ocorre, existem 80000 quadros de +recursão+ na pilha!
(((rastreamento de pilha)))(((StackOverflowError)))((("Erro", "Core", "StackOverflowError", see="StackOverflowError")))

> Se você encontrar uma recursão infinita por acidente, revise sua função para confirmar se há um caso base que não faz uma chamada recursiva. E se houver um caso base, verifique se você está garantindo para alcançá-lo.

Reviewer 2: ```suggestion
Se você encontrar uma recursão infinita por acidente, revise sua função para confirmar se há um caso base que não faz uma chamada recursiva. E se houver um caso base, verifique que você está garantido a alcançá-lo.
```

> === Entradas de Teclado

Reviewer 2: ```suggestion
=== Entradas do Teclado
```

Os programas que escrevemos até agora não aceitam nenhuma entrada do usuário. Eles apenas fazem a mesma coisa todas as vezes.

Julia fornece uma função interna chamada +readline+ que interrompe o programa e aguarda o usuário digitar algo. Quando o usuário pressiona +RETURN+ ou +ENTER+, o programa é retomado e +readline+ retorna o que o usuário digitou como uma sequência de caracteres.

> (((readline)))((("function", "Base", "readline", see="readline")))

Reviewer 2: ```suggestion
(((readline)))((("função, "Base", "readline", see="readline")))
```

[source,jlcon]
----
julia> texto = readline()
O que você está esperando?
"O que você está esperando?"
----

Antes de receber informações do usuário, é uma boa ideia imprimir um prompt informando ao usuário o que digitar:
(((prompt)))

[source,jlcon]
----
julia> print("Quem és tu? "); readline()
Quem és tu? Sou Hermanoteu da Pentescopéia, irmão da Micalatéia.
"Sou Hermanoteu da Pentescopéia, irmão da Micalatéia."
----

> Um ponto e vírgula +;+ permite colocar multiplas atribuições na mesma linha. No REPL apenas a última atribuição retornará seu valor.

Reviewer: ```suggestion
Um ponto e vírgula +;+ permite colocar múltiplos comandos na mesma linha. No REPL apenas o último comando retornará seu valor.
```

(((;)))

Se você espera que o usuário digite um número inteiro, tente converter o valor de retorno para +Int64+:

[source,jlcon]
----

> julia> println("Qual é a velocidade da velocidade do ar de uma andorinha sem carga? "); speed = readline()
> Qual é a velocidade da velocidade do ar de uma andorinha sem carga?

Reviewer: ```suggestion
julia> println("Qual é a velocidade de vôo de uma andorinha sem carga?"); velocidade = readline()
Qual é a velocidade de vôo de uma andorinha sem carga?
```

42
"42"
julia> parse(Int64, speed)
42
----

Mas se o usuário digitar algo diferente de uma sequência de dígitos, você receberá um erro:
(((analisar)))

[source,jlcon]
----

> julia> println("Qual é a velocidade da velocidade do ar de uma andorinha sem carga? "); velocidade = readline()
> Qual é a velocidade da velocidade do ar de uma andorinha sem carga?
> Como assim, uma andorinha africana ou européia?
> "Como assim, uma andorinha africana ou européia?"

Reviewer: ```suggestion
julia> println("Qual é a velocidade de vôo de uma andorinha sem carga?"); velocidade = readline()
Qual é a velocidade de vôo de uma andorinha sem carga?
Como assim, uma andorinha africana ou européia?
"Como assim, uma andorinha africana ou européia?"
```

julia> parse(Int64, speed)
ERROR: ArgumentError: invalid base 10 digit 'C' in "Como assim, uma andorinha africana ou européia?"
[...]
----

Veremos como lidar com esse tipo de erro posteriormente.

> (((Erro de Argumento)))

Reviewer 2: ```suggestion
(((ArgumentError)))
```

=== Debugando

> Quando ocorre um erro de sintaxe ou de tempo de execução, a mensagem de erro contém muitas informações, mas pode ser avassaladora. As partes mais úteis são geralmente:

Reviewer 2: ```suggestion
Quando um erro de sintaxe ou de tempo de execução ocorrer, a mensagem de erro contém muitas informações, mas ela pode ser avassaladora. As partes mais úteis são geralmente:
```

(((debugando)))

* Que tipo de erro foi, e

* Onde ocorreu.

> Os erros de sintaxe geralmente são fáceis de encontrar, mas existem algumas dicas. Em geral, as mensagens de erro indicam onde o problema foi descoberto, mas o verdadeiro erro pode estar antes no código, às vezes em uma linha anterior.

Reviewer: ```suggestion
Os erros de sintaxe geralmente são fáceis de encontrar, mas existem algumas ressalvas. Em geral, as mensagens de erro indicam onde o problema foi descoberto, mas o verdadeiro erro pode estar antes no código, às vezes em uma linha anterior.
```

O mesmo vale para erros de tempo de execução. Suponha que você esteja tentando calcular uma taxa de sinal/ruído em decibéis. A fórmula é

[latexmath]
++++
\begin{equation}
{SNR_{\mathrm{db}} = 10 \log_{10} \frac{P_{\mathrm{sinal}}}{P_{\mathrm{ruido}}}\ .}
\end{equation}
++++

> Em Julia, você pode escrever algo como isto:

Reviewer 2: ```suggestion
Em Julia, você pode escrever algo como:
```

[source,julia]
----
potência_do_sinal = 9
potência_do_ruido = 10
razão = potência_do_sinal ÷ potência_do_ruido
decibéis = 10 * log10(razão)
print(decibeis)
----

E você obtém:

[source,julia]
----
-Inf
----

Esse não era um resultado que você estava esperando.

> Para encontrar o erro, pode ser útil imprimir o valor da razão, que acaba sendo 0. O problema está na linha 3, que usa a Divisão do piso em vez da Divisão do ponto flutuante.

Reviewer: ```suggestion
Para encontrar o erro, pode ser útil imprimir o valor da razão, que acaba sendo 0. O problema está na linha 3, que usa a divisão do piso em vez da divisão do ponto flutuante.
```

[WARNING]
====
Você deve ler atentamente as mensagens de erro, mas não presuma que tudo o que elas dizem está correto.
====

=== Glossário

> Divisão Inteira::

Reviewer: ```suggestion
divisão inteira::
```

Um operador, denotado +÷+, que divide dois números e arredonda para baixo (em direção ao infinito negativo) para um número inteiro.

> (((Divisão Inteira)))

Reviewer: ```suggestion
(((divisão inteira)))
```

operador módulo::
Um operador, indicado com um sinal de porcentagem (%), que trabalha com números inteiros e retorna o restante quando um número é dividido por outro.
(((operador módulo)))

expressão booleana::
Uma expressão cujo os valores são ou +true+ ou +false+.
(((expressão booleana)))

operador relacional::
Um dos operadores que compara operandos: +==+, +≠+ (+!=+), +>+, +<+, +≥+ (+>=+), e +≤+ (+pass:[&lt;=]+).
(((operador relacional)))

operador lógico::
Um dos operadores que combina expressões booleanas: +&&+ (e), +||+ (ou), e +!+ (não).
(((operador lógico)))

> atribuição condicional::

Reviewer: ```suggestion
declaração condicional::
```

> Uma atribuição que controla o fluxo de execução dependendo de alguma condição.

Reviewer: ```suggestion
Uma declaração que controla o fluxo de execução dependendo de alguma condição.
```

> (((atribuição condicional)))

Reviewer: ```suggestion
(((declaração condicional)))
```

condição::

> A expressão boleana em uma atribuição condicional que determina qual ramo executará.

Reviewer: ```suggestion
A expressão boleana em uma declaração condicional que determina qual ramo executará.
```

(((condição)))

> atribuição composta::
> Uma atribuição que consiste em um cabeçalho e um corpo. O corpo é terminado com a palavra-chave +end+.
> (((atribuição composta)))

Reviewer: ```suggestion
declaração composta::
Uma declaração que consiste em um cabeçalho e um corpo. O corpo é terminado com a palavra-chave +end+.
(((declaração composta)))
```

ramos::
Uma das sequências alternativas de atribuições em uma atribuição condicional.
(((ramos)))

atribuição encadeada::
Uma atribuição condicional com uma série de ramos alternativos.
(((atribuição encadeada)))

condicional aninhada::
Uma atribuição condicional que aparece em um dos ramos de outra atribuição condicional.
(((condicional aninhada)))

atribuição de retorno::

> Uma atribuição que faz com que uma função pare de executar imediatamente e retorne para quem chamou.

Reviewer 2: ```suggestion
Uma atribuição que faz com que uma função pare de executar imediatamente e retorne para quem a chamou.
```

> (((atribuição de retorno)))

Reviewer: Mudar todas essas atribuições

recursão::
O processo de chamar a função que está sendo executada.
(((recursão)))

caso base::

> Uma ramo condicional de uma função recursiva que não faz um chamado recursivo.

Reviewer 2: ```suggestion
Uma ramo condicional de uma função recursiva que não faz um chamado recursiva.
```

(((caso base)))

recursão infinita::
Uma recursão que não tem um caso base, ou nunca atinge ela. Eventualmente, uma recursão infinita causa um erro de tempo de execução.
(((recursão infinita)))

=== Exercícios

[[ex05-1]]
===== Exercício 5-2

> A função +tempo+ retorna o tempo médio atual de Greenwich em segundos desde “a época”, que é um tempo arbitrário usado como ponto de referência. Nos sistemas UNIX, a época é 1 de janeiro de 1970.
> (((tempo)))((("função", "Base", "hora", see="hora")))

Reviewer: ```suggestion
A função +time+ retorna o tempo médio atual de Greenwich em segundos desde “a época”, que é um tempo arbitrário usado como ponto de referência. Nos sistemas UNIX, a época é 1 de janeiro de 1970.
(((time)))((("função", "Base", "time", see="time")))
```

[source,@julia-repl]
----

> tempo()

Reviewer: ```suggestion
time()
```

julia> recursão()
ERROR: StackOverflowError:
Stacktrace:
 [1] recursão() at ./REPL[1]:2 (repeats 80000 times)
----

Esse rastreamento de pilha é um pouco maior do que vimos no capítulo anterior. Quando o erro ocorre, existem 80000 quadros de +recursão+ na pilha!
(((rastreamento de pilha)))(((StackOverflowError)))((("Erro", "Core", "StackOverflowError", see="StackOverflowError")))

Se você encontrar uma recursão infinita por acidente, revise a sua função para confirmar se há um caso base que não faz uma chamada recursiva. E se houver, verifique que você está garantindo o alcance do caso base.

=== Entradas de Teclado

Os programas que escrevemos até agora não aceitam nenhuma entrada do usuário. Eles apenas fazem a mesma coisa toda hora.

> O Julia fornece uma função interna chamada +readline+ que interrompe o programa e aguarda o usuário digitar algo. Quando o usuário pressiona +return+ ou +ENTER+, o programa é retomado e +readline+ retorna o que o usuário digitou como uma sequência de caracteres.

Reviewer: ```suggestion
O Julia fornece uma função interna chamada +readline+ que interrompe o programa e aguarda o usuário digitar algo. Quando o usuário pressiona +RETURN+ ou +ENTER+, o programa é retomado e +readline+ retorna o que o usuário digitou como uma sequência de caracteres.
```

\end{equation}
++++

para qualquer valor de latexmath:[\(n\)] maior que 2.

. Escreva uma função chamada +fermat+ que coleta quatro parâmetros — +a+, +b+, +c+ e +n+ — e verifica se o Teorema de Fermat é valido. Se +n+ é maior que 2 e +pass:[a^n + b^n == c^n]+ o programa deve imprimir, "Oloco, Fermat estava errado!" caso contrário o programa deve imprimir, "Não, isso não funciona."

> (((fermat)))((("function", "definido pelo programador", "fermat", see="fermat")))

Reviewer 2: função

. Escreva uma função que solicite ao usuário que insira valores para +a+, +b+, +c+ e +n+, converta-os em números inteiros e use +fermat+ para verificar se violam o teorema de Fermat.

[[ex05-3]]
===== Exercise 5-4

Se você receber três gravetos, poderá ou não ser capaz de organizá-los em um triângulo. Por exemplo, se um dos gravetos tiver 12 centímetros de comprimento e os outros dois um centímetro, você não conseguirá colocar os gravetos curtos no meio. Para três comprimentos, há um teste simples para verificar se é possível formar um triângulo:

[TIP]
====
Se qualquer um dos três comprimentos for maior que a soma dos outros dois, não será possível formar um triângulo. Caso contrário, você pode. (Se a soma de dois comprimentos for igual ao terceiro, eles formarão o que é chamado de triângulo "degenerado".)
====

> . Escreva uma função chamada +triangulo+ que aceite três números inteiros como argumentos e imprima “Sim” ou “Não”, dependendo de você poder ou não formar um triângulo a partir de gravetos com os comprimentos especificados.

Reviewer 2: "dependendo de você poder ou não formar ... " ->" dependendo da possibilidade de formar ou não ... "

> . Escreva uma função chamada +triangulo+ que aceite três números inteiros como argumentos e imprima “Sim” ou “Não”, dependendo de você poder ou não formar um triângulo a partir de gravetos com os comprimentos especificados.

Reviewer 2: "triangulo" -> "triângulo", o julia aceita funções com acento, então não tem problema. Trocar também pros outros casos

(((triangulo)))((("função", "definido pelo programador", "é triângulo", see="é triângulo")))

. Escreva uma função que solicite ao usuário a inserção de três comprimentos de gravetos, os converta em números inteiros e use +triangulo+ para verificar se os gravetos com os comprimentos especificados podem formar um triângulo.

[[ex05-4]]
===== Exercício 5-5

> Qual é a saída do seguinte programa? Desenhe um diagrama de pilha que mostre o estado do programa quando imprimir o resultado.

Reviewer 2: "quando imprimir" -> "quando ele imprimir"

. Escreva uma sequência de documentos que explique tudo o que alguém precisa saber para usar essa função (e nada mais).

Os exercícios a seguir usam o módulo +ThinkJulia+, descrito no <<chap04>>:

[[ex05-5]]
===== Exercício 5-6

Leia a função a seguir e veja se você consegue descobrir o que ela faz (veja os exemplos em <<chap04>>). Em seguida, execute-o e veja se você acertou.

[source,julia]
----
function desenhe(t, comprimento, n)
    if n == 0
        return
    end
    angulo = 50
    forward(t, comprimento*n)
    turn(t, -angulo)
    draw(t, comprimento, n-1)
    turn(t, 2*angle)
    draw(t, comprimento, n-1)
    turn(t, -angulo)
    forward(t, -comprimento*n)
end
----

[[ex05-6]]
===== Exercício 5-7

[[fig05-2]]
.A curva de Koch
image::images/fig52.svg[]

A curva de Koch é um fractal que se parece com <<fig05-2>>. Para desenhar uma curva de Koch com comprimento latexmath:[\(x\)], tudo que você precisa fazer é:
(((curva de Koch)))(((fractal)))

. Desenhe uma curva de Koch com comprimento latexmath:[\(\frac{x}{3}\)].

. Gire 60° para esquerda

. Desenhe uma curva de Koch com comprimento latexmath:[\(\frac{x}{3}\)].

> . Fire 120° para direita.

Reviewer 2: vire

Author: Substituido por Gire conforme padrão anterior

\end{equation}
++++

para qualquer valor de latexmath:[\(n\)] maior que 2.

. Escreva uma função chamada +fermat+ que coleta quatro parâmetros — +a+, +b+, +c+ e +n+ — e verifica se o Teorema de Fermat é valido. Se +n+ é maior que 2 e +pass:[a^n + b^n == c^n]+ o programa deve imprimir, "Oloco, Fermat estava errado!" caso contrário o programa deve imprimir, "Não, isso não funciona."
(((fermat)))((("função", "definido pelo programador", "fermat", see="fermat")))

> . Escreva uma função que solicite ao usuário que insira valores para +a+, +b+, +c+ e +n+, converta-os em números inteiros e use +fermat+ para verificar se violam o teorema de Fermat.

Reviewer 2: ```suggestion
. Escreva uma função que solicite ao usuário que insira valores para +a+, +b+, +c+ e +n+, converta-os em números inteiros e use +fermat+ para verificar se eles violam o teorema de Fermat.
```

[[ex05-3]]
===== Exercise 5-4

> Se você receber três gravetos, poderá ou não ser capaz de organizá-los em um triângulo. Por exemplo, se um dos gravetos tiver 12 centímetros de comprimento e os outros dois um centímetro, você não conseguirá colocar os gravetos curtos no meio. Para três comprimentos, há um teste simples para verificar se é possível formar um triângulo:

Reviewer: ```suggestion
Se você receber três gravetos, poderá ou não ser capaz de organizá-los em um triângulo. Por exemplo, se um dos gravetos tiver 12 centímetros de comprimento e os outros dois um centímetro, você não conseguirá fazer com que os gravetos curtos se toquem no meio. Para três comprimentos, há um teste simples para verificar se é possível formar um triângulo:
```

[TIP]
====
Se qualquer um dos três comprimentos for maior que a soma dos outros dois, não será possível formar um triângulo. Caso contrário, você pode. (Se a soma de dois comprimentos for igual ao terceiro, eles formarão o que é chamado de triângulo "degenerado".)
====

> . Escreva uma função chamada +triângulo+ que aceite três números inteiros como argumentos e imprima “Sim” ou “Não”, dependendo da possibilidade de formar ou não um triângulo a partir de gravetos com os comprimentos especificados.

Reviewer 2: ```suggestion
. Escreva uma função chamada +é_triângulo+ que aceite três números inteiros como argumentos e imprima “Sim” ou “Não”, dependendo da possibilidade de formar ou não um triângulo a partir de gravetos com os comprimentos especificados.
```

> (((triângulo)))((("função", "definido pelo programador", "é triângulo", see="é triângulo")))

Reviewer 2: ```suggestion
(((triângulo)))((("função", "definido pelo programador", "é_triângulo", see="é_triângulo")))
```

> . Escreva uma função que solicite ao usuário a inserção de três comprimentos de gravetos, os converta em números inteiros e use +triangulo+ para verificar se os gravetos com os comprimentos especificados podem formar um triângulo.

Reviewer 2: ```suggestion
. Escreva uma função que solicite ao usuário a inserção de três comprimentos de gravetos, os converta em números inteiros e use +é_triângulo+ para verificar se os gravetos com os comprimentos especificados podem formar um triângulo.
```

[[ex05-4]]
===== Exercício 5-5

Qual é a saída do seguinte programa? Desenhe um diagrama de pilha que mostre o estado do programa quando ele imprimir o resultado.
(((diagrama de pilha)))(((recursão)))

[source,julia]
----
function recursão(n, s)
    if n == 0
        println(s)
    else
        recursão(n-1, n+s)
    end
end

recursão(3, 0)
----

. O que aconteceria se você chamasse essa função assim: +recursão(-1, 0)+?

. Escreva uma sequência de documentos que explique tudo o que alguém precisa saber para usar essa função (e nada mais).

Os exercícios a seguir usam o módulo +ThinkJulia+, descrito no <<chap04>>:

[[ex05-5]]
===== Exercício 5-6

Leia a função a seguir e veja se você consegue descobrir o que ela faz (veja os exemplos em <<chap04>>). Em seguida, execute-o e veja se você acertou.

[source,julia]
----
function desenhe(t, comprimento, n)
    if n == 0
        return
    end
    angulo = 50
    forward(t, comprimento*n)
    turn(t, -angulo)
    draw(t, comprimento, n-1)

>     turn(t, 2*angle)

Reviewer: !

>     ângulo = 50
>     avance(t, comprimento*n)
>     gire(t, -ângulo)
>     desenhe(t, comprimento, n-1)
>     gire(t, 2*angle)
>     desenhe(t, comprimento, n-1)
>     gire(t, -ângulo)
>     avance(t, -comprimento*n)

Reviewer: ```suggestion
    ângulo = 50
    forward(t, comprimento*n)
    turn(t, -ângulo)
    draw(t, comprimento, n-1)
    turn(t, 2*angle)
    draw(t, comprimento, n-1)
    turn(t, -ângulo)
    forward(t, -comprimento*n)
```

end
----

[[ex05-6]]
===== Exercício 5-7

[[fig05-2]]
.A curva de Koch
image::images/fig52.svg[]

A curva de Koch é um fractal que se parece com <<fig05-2>>. Para desenhar uma curva de Koch com comprimento latexmath:[\(x\)], tudo que você precisa fazer é:
(((curva de Koch)))(((fractal)))

. Desenhe uma curva de Koch com comprimento latexmath:[\(\frac{x}{3}\)].

. Gire 60° para esquerda

. Desenhe uma curva de Koch com comprimento latexmath:[\(\frac{x}{3}\)].

. Gire 120° para direita.

. Desenhe uma curva de Koch com comprimento latexmath:[\(\frac{x}{3}\)].

. Gire 60° para esquerda.

. Desenhe uma curva de Koch com comprimento latexmath:[\(\frac{x}{3}\)].

A exceção é se latexmath:[\(x\)] for menor que 3: neste caso, você só desenha uma linha reta de comprimento latexmath:[\(x\)].

. Escreva uma função chamada +koch+ que considere uma tartaruga e um comprimento como parâmetros e que use a tartaruga para desenhar uma curva de Koch com o comprimento especificado.

> (((koch)))((("function", "definido pelo programador", "koch", see="koch")))

Reviewer 2: ```suggestion
(((koch)))((("função", "definido pelo programador", "koch", see="koch")))
```

> . Escreva uma função chamada + floco de neve + que desenhe três curvas de Koch para fazer o contorno de um floco de neve.

Reviewer 2: ```suggestion
. Escreva uma função chamada + floco_de_neve + que desenhe três curvas de Koch para fazer o contorno de um floco de neve.
```

> (((floco de neve)))((("function", "definido pelo programador", "floco de neve", see="floco de neve")))

Reviewer 2: ```suggestion
(((floco_de_neve)))((("function", "definido pelo programador", "floco_de_neve", see="floco _de_neve")))
```

> . A curva de Koch pode ser generalizada de várias maneiras. Veja https://en.wikipedia.org/wiki/Koch_snowflake para exemplos e implemente o seu favorito.

Reviewer: ```suggestion
. A curva de Koch pode ser generalizada de várias maneiras. Veja https://en.wikipedia.org/wiki/Koch_snowflake para exemplos e implemente o seu favorito.

```
